File: app/src/main/java/com/mkulesh/onpc/iscp/messages/CdPlayerOperationCommandMsg.java
/*
 * Copyright (C) 2018. Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

package com.mkulesh.onpc.iscp.messages;

import com.mkulesh.onpc.R;
import com.mkulesh.onpc.iscp.EISCPMessage;
import com.mkulesh.onpc.iscp.ISCPMessage;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.StringRes;

/*
 * CD Player Operation Command
 */
public class CdPlayerOperationCommandMsg extends ISCPMessage
{
    public final static String CODE = "CCD";

    // Controls that allow the control of build-in CD player via CCD command
    public final static String CONTROL_CD_INT1 = "CD Control";
    public final static String CONTROL_CD_INT2 = "CD Control(NewRemote)";

    public enum Command implements StringParameterIf
    {
        POWER(R.string.cd_cmd_power, R.drawable.menu_power_standby),
        TRACK(R.string.cd_cmd_track),
        PLAY(R.string.cd_cmd_play, R.drawable.cmd_play),
        STOP(R.string.cd_cmd_stop, R.drawable.cmd_stop),
        PAUSE(R.string.cd_cmd_pause, R.drawable.cmd_pause),
        SKIP_F(R.string.cd_cmd_skip_f, R.drawable.cmd_next, "SKIP.F"),
        SKIP_R(R.string.cd_cmd_skip_r, R.drawable.cmd_previous, "SKIP.R"),
        MEMORY(R.string.cd_cmd_memory),
        CLEAR(R.string.cd_cmd_clear),
        REPEAT(R.string.cd_cmd_repeat, R.drawable.repeat_all),
        RANDOM(R.string.cd_cmd_random, R.drawable.cmd_random),
        DISP(R.string.cd_cmd_disp),
        D_MODE(R.string.cd_cmd_d_mode),
        FF(R.string.cd_cmd_ff),
        REW(R.string.cd_cmd_rew),
        OP_CL(R.string.cd_cmd_op_cl, R.drawable.cd_eject, "OP/CL"),
        NUMBER_1(R.string.cd_cmd_number_1, "1"),
        NUMBER_2(R.string.cd_cmd_number_2, "2"),
        NUMBER_3(R.string.cd_cmd_number_3, "3"),
        NUMBER_4(R.string.cd_cmd_number_4, "4"),
        NUMBER_5(R.string.cd_cmd_number_5, "5"),
        NUMBER_6(R.string.cd_cmd_number_6, "6"),
        NUMBER_7(R.string.cd_cmd_number_7, "7"),
        NUMBER_8(R.string.cd_cmd_number_8, "8"),
        NUMBER_9(R.string.cd_cmd_number_9, "9"),
        NUMBER_0(R.string.cd_cmd_number_0, "0"),
        NUMBER_10(R.string.cd_cmd_number_10, "10"),
        NUMBER_GREATER_10(R.string.cd_cmd_number_greater_10, "+10"),
        DISC_F(R.string.cd_cmd_disc_f),
        DISC_R(R.string.cd_cmd_disc_r),
        DISC1(R.string.cd_cmd_disc1),
        DISC2(R.string.cd_cmd_disc2),
        DISC3(R.string.cd_cmd_disc3),
        DISC4(R.string.cd_cmd_disc4),
        DISC5(R.string.cd_cmd_disc5),
        DISC6(R.string.cd_cmd_disc6);

        @StringRes
        final int descriptionId;

        @DrawableRes
        final int imageId;
        final String cmd;

        Command(@StringRes final int descriptionId)
        {
            this.descriptionId = descriptionId;
            this.imageId = R.drawable.media_item_unknown;
            this.cmd = null;
        }

        Command(@StringRes final int descriptionId, @DrawableRes final int imageId)
        {
            this.descriptionId = descriptionId;
            this.imageId = imageId;
            this.cmd = null;
        }

        Command(@StringRes final int descriptionId, final String cmd)
        {
            this.descriptionId = descriptionId;
            this.imageId = R.drawable.media_item_unknown;
            this.cmd = cmd;
        }

        Command(@StringRes final int descriptionId, @DrawableRes final int imageId, final String cmd)
        {
            this.descriptionId = descriptionId;
            this.imageId = imageId;
            this.cmd = cmd;
        }

        public String getCode()
        {
            return toString();
        }

        @StringRes
        public int getDescriptionId()
        {
            return descriptionId;
        }

        @DrawableRes
        public int getImageId()
        {
            return imageId;
        }

        String getCmd()
        {
            return cmd != null ? cmd : getCode();
        }
    }

    private final Command command;

    CdPlayerOperationCommandMsg(EISCPMessage raw) throws Exception
    {
        super(raw);
        this.command = (Command) searchParameter(data, Command.values(), null);
    }

    public CdPlayerOperationCommandMsg(final String command)
    {
        super(0, null);
        this.command = (Command) searchParameter(command, Command.values(), null);
    }

    public Command getCommand()
    {
        return command;
    }

    @NonNull
    @Override
    public String toString()
    {
        return CODE + "[" + data
                + "; CMD=" + (command == null ? "null" : command.getCmd())
                + "]";
    }

    @Override
    public EISCPMessage getCmdMsg()
    {
        return command == null ? null : new EISCPMessage(CODE, command.getCmd());
    }

    @Override
    public boolean hasImpactOnMediaList()
    {
        return false;
    }

    @NonNull
    public static String convertOpCommand(@NonNull final String opCommand)
    {
        switch (opCommand)
        {
            case "TRDN":
                return Command.SKIP_R.getCode();
            case "TRUP":
                return Command.SKIP_F.getCode();
            default:
                return opCommand;
        }
    }
}

File: lib/iscp/messages/CdPlayerOperationCommandMsg.dart
/*
 * Enhanced Controller for Onkyo and Pioneer Pro
 * Copyright (C) 2019-2021 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

import "../../constants/Drawables.dart";
import "../../constants/Strings.dart";
import "EnumParameterMsg.dart";

enum CdPlayerOperationCommand
{
    UNDEFINED,
    POWER,
    TRACK,
    PLAY,
    STOP,
    PAUSE,
    SKIP_F,
    SKIP_R,
    MEMORY,
    CLEAR,
    REPEAT,
    RANDOM,
    DISP,
    D_MODE,
    FF,
    REW,
    OP_CL,
    NUMBER_1,
    NUMBER_2,
    NUMBER_3,
    NUMBER_4,
    NUMBER_5,
    NUMBER_6,
    NUMBER_7,
    NUMBER_8,
    NUMBER_9,
    NUMBER_0,
    NUMBER_10,
    NUMBER_GREATER_10,
    DISC_F,
    DISC_R,
    DISC1,
    DISC2,
    DISC3,
    DISC4,
    DISC5,
    DISC6
}

/*
 * CD Player Operation Command
 */
class CdPlayerOperationCommandMsg extends EnumParameterMsg<CdPlayerOperationCommand>
{
    static const String CODE = "CCD";

    // Controls that allow the control of build-in CD player via CCD command
    static const String CONTROL_CD_INT1 = "CD Control";
    static const String CONTROL_CD_INT2 = "CD Control(NewRemote)";

    static const ExtEnum<CdPlayerOperationCommand> ValueEnum = ExtEnum<CdPlayerOperationCommand>([
        EnumItem.code(CdPlayerOperationCommand.UNDEFINED, "N/A",
            descr: Strings.dashed_string, defValue: true),
        EnumItem.code(CdPlayerOperationCommand.POWER, "POWER",
            descrList: Strings.l_cd_cmd_power, icon: Drawables.menu_power_standby),
        EnumItem.code(CdPlayerOperationCommand.TRACK, "TRACK",
            descrList: Strings.l_cd_cmd_track),
        EnumItem.code(CdPlayerOperationCommand.PLAY, "PLAY",
            descrList: Strings.l_cd_cmd_play, icon: Drawables.cmd_play),
        EnumItem.code(CdPlayerOperationCommand.STOP, "STOP",
            descrList: Strings.l_cd_cmd_stop, icon: Drawables.cmd_stop),
        EnumItem.code(CdPlayerOperationCommand.PAUSE, "PAUSE",
            descrList: Strings.l_cd_cmd_pause, icon: Drawables.cmd_pause),
        EnumItem.code(CdPlayerOperationCommand.SKIP_F, "SKIP.F",
            descrList: Strings.l_cd_cmd_skip_f, icon: Drawables.cmd_next),
        EnumItem.code(CdPlayerOperationCommand.SKIP_R, "SKIP.R",
            descrList: Strings.l_cd_cmd_skip_r, icon: Drawables.cmd_previous),
        EnumItem.code(CdPlayerOperationCommand.MEMORY, "MEMORY",
            descrList: Strings.l_cd_cmd_memory),
        EnumItem.code(CdPlayerOperationCommand.CLEAR, "CLEAR", name: "C",
            descrList: Strings.l_cd_cmd_clear, icon: Drawables.numeric_clean),
        EnumItem.code(CdPlayerOperationCommand.REPEAT, "REPEAT",
            descrList: Strings.l_cd_cmd_repeat, icon: Drawables.repeat_all),
        EnumItem.code(CdPlayerOperationCommand.RANDOM, "RANDOM",
            descrList: Strings.l_cd_cmd_random, icon: Drawables.cmd_random),
        EnumItem.code(CdPlayerOperationCommand.DISP, "DISP",
            descrList: Strings.l_cd_cmd_disp),
        EnumItem.code(CdPlayerOperationCommand.D_MODE, "D.MODE",
            descrList: Strings.l_cd_cmd_d_mode),
        EnumItem.code(CdPlayerOperationCommand.FF, "FF",
            descrList: Strings.l_cd_cmd_ff),
        EnumItem.code(CdPlayerOperationCommand.REW, "REW",
            descrList: Strings.l_cd_cmd_rew),
        EnumItem.code(CdPlayerOperationCommand.OP_CL, "OP/CL",
            descrList: Strings.l_cd_cmd_op_cl, icon: Drawables.cd_eject),
        EnumItem.code(CdPlayerOperationCommand.NUMBER_1, "1", name: "1",
            descrList: Strings.l_cd_cmd_number_1, icon: Drawables.numeric_1),
        EnumItem.code(CdPlayerOperationCommand.NUMBER_2, "2", name: "2",
            descrList: Strings.l_cd_cmd_number_2, icon: Drawables.numeric_2),
        EnumItem.code(CdPlayerOperationCommand.NUMBER_3, "3", name: "3",
            descrList: Strings.l_cd_cmd_number_3, icon: Drawables.numeric_3),
        EnumItem.code(CdPlayerOperationCommand.NUMBER_4, "4", name: "4",
            descrList: Strings.l_cd_cmd_number_4, icon: Drawables.numeric_4),
        EnumItem.code(CdPlayerOperationCommand.NUMBER_5, "5", name: "5",
            descrList: Strings.l_cd_cmd_number_5, icon: Drawables.numeric_5),
        EnumItem.code(CdPlayerOperationCommand.NUMBER_6, "6", name: "6",
            descrList: Strings.l_cd_cmd_number_6, icon: Drawables.numeric_6),
        EnumItem.code(CdPlayerOperationCommand.NUMBER_7, "7", name: "7",
            descrList: Strings.l_cd_cmd_number_7, icon: Drawables.numeric_7),
        EnumItem.code(CdPlayerOperationCommand.NUMBER_8, "8", name: "8",
            descrList: Strings.l_cd_cmd_number_8, icon: Drawables.numeric_8),
        EnumItem.code(CdPlayerOperationCommand.NUMBER_9, "9", name: "9",
            descrList: Strings.l_cd_cmd_number_9, icon: Drawables.numeric_9),
        EnumItem.code(CdPlayerOperationCommand.NUMBER_0, "0", name: "0",
            descrList: Strings.l_cd_cmd_number_0, icon: Drawables.numeric_0),
        EnumItem.code(CdPlayerOperationCommand.NUMBER_10, "10", name: "10",
            descrList: Strings.l_cd_cmd_number_10),
        EnumItem.code(CdPlayerOperationCommand.NUMBER_GREATER_10, "+10", name: ">9",
            descrList: Strings.l_cd_cmd_number_greater_10, icon: Drawables.numeric_greater_9),
        EnumItem.code(CdPlayerOperationCommand.DISC_F, "DISC.F",
            descrList: Strings.l_cd_cmd_disc_f),
        EnumItem.code(CdPlayerOperationCommand.DISC_R, "DISC.R",
            descrList: Strings.l_cd_cmd_disc_r),
        EnumItem.code(CdPlayerOperationCommand.DISC1, "DISC1",
            descrList: Strings.l_cd_cmd_disc1),
        EnumItem.code(CdPlayerOperationCommand.DISC2, "DISC2",
            descrList: Strings.l_cd_cmd_disc2),
        EnumItem.code(CdPlayerOperationCommand.DISC3, "DISC3",
            descrList: Strings.l_cd_cmd_disc3),
        EnumItem.code(CdPlayerOperationCommand.DISC4, "DISC4",
            descrList: Strings.l_cd_cmd_disc4),
        EnumItem.code(CdPlayerOperationCommand.DISC5, "DISC5",
            descrList: Strings.l_cd_cmd_disc5),
        EnumItem.code(CdPlayerOperationCommand.DISC6, "DISC6",
            descrList: Strings.l_cd_cmd_disc6)
    ]);

    CdPlayerOperationCommandMsg.output(CdPlayerOperationCommand v) : super.output(CODE, v, ValueEnum);

    @override
    bool hasImpactOnMediaList()
    {
        return false;
    }
}
